pkg/kv/kvserver/deleted_external_sstable_test.go

```diff
@@ -868,3 +870,132 @@ func TestGeneralOperationsWorkAsExpectedOnDeletedExternalSST(t *testing.T) {
+	const externURI = "nodelocal://1/external-files"
+
+	ctx := context.Background()
+	etc := externalSSTTestCluster{}
+
+	// Create a test cluster with the following ranges: [a, d), [d, g), [g, z).
+	etc.testSetup(t)
+	defer etc.tc.Stopper().Stop(ctx)
+
+	externalStorage, err := etc.createExternalStorage(ctx, externURI)
+	require.NoError(t, err)
+
+	deletedStartKey := roachpb.Key("d")
+	deletedEndKey := roachpb.Key("k")
+	firstStartChar := deletedStartKey[0]
+	firstEndChar := deletedEndKey[0]
+
+	require.NoError(t, etc.createExternalSSTableFile(t, ctx, externalStorage,
+		"file1.sst", firstStartChar, firstEndChar))
+	require.NoError(t, etc.linkExternalSSTableToFile(ctx, deletedStartKey,
+		deletedEndKey, externURI, "file1.sst"))
+
+	// Create one range feed per range.
+	rfStream1, rfErr1 := etc.createRangeFeed(t, roachpb.Key("a"), roachpb.Key("d"))
+	defer rfStream1.Cancel()
+
+	rfStream2, rfErr2 := etc.createRangeFeed(t, roachpb.Key("d"), roachpb.Key("g"))
+	defer rfStream2.Cancel()
+
+	rfStream3, rfErr3 := etc.createRangeFeed(t, roachpb.Key("g"), roachpb.Key("z"))
+	defer rfStream3.Cancel()
+
+	waitForCheckpoint := func(rfErr chan error, stream *testStream) {
+		require.Eventually(t, func() bool {
+			if len(rfErr) > 0 {
+				require.Fail(t, "unexpected rangefeed error", "%v", <-rfErr)
+			}
+			events := stream.Events()
+			for _, event := range events {
+				require.NotNil(t, event.Checkpoint, "received non-checkpoint event: %v", event)
+			}
+			return len(events) > 0
+		}, 5*time.Second, 100*time.Millisecond)
+	}
+
+	// Wait for the RangeFeeds to receive the initial checkpoint. This is to avoid
+	// the RangeFeed racing with the deleted span, which would cause the
+	// RangeFeed to fail to start.
+	waitForCheckpoint(rfErr1, rfStream1)
+	waitForCheckpoint(rfErr2, rfStream2)
+	waitForCheckpoint(rfErr3, rfStream3)
+
+	// Perform some data operations, and delete the external SSTable file.
+	pendingTxn1, pendingTxn2, err := etc.writeIntents(ctx, etc.db)
+	require.NoError(t, err)
+
+	require.NoError(t, externalStorage.Delete(ctx, "file1.sst"))
+
+	require.NoError(t, pendingTxn1.Commit(ctx))
+	require.NoError(t, pendingTxn2.Commit(ctx))
+	require.NoError(t, etc.putHelper(ctx, roachpb.Key("a")))
+	// Requests to the deleted key span should fail.
+	etc.requireNotFoundError(t, etc.putHelper(ctx, roachpb.Key("e-15000")))
+	etc.requireNotFoundError(t, etc.putHelper(ctx, roachpb.Key("h-15000")))
+	require.NoError(t, etc.putHelper(ctx, roachpb.Key("y")))
+	require.NoError(t, etc.deleteRangeHelper(ctx, roachpb.Key("a"), roachpb.Key("b")))
+	require.NoError(t, etc.deleteRangeHelper(ctx, roachpb.Key("k"), roachpb.Key("l")))
+
+	// Before running Excise, the RangeFeeds should still be running.
+	require.Equal(t, 0, len(rfErr1))
+	require.Equal(t, 0, len(rfErr2))
+	require.Equal(t, 0, len(rfErr3))
+
+	waitForExciseError := func(rfErr chan error) {
+		testutils.SucceedsSoon(t, func() error {
+			if len(rfErr) == 0 {
```

Comment:
I would have written this as follows:

```go
select {
  case err := <-rfErr:
    ...
  default:
    return errors.Errorf("no error was found on the range feed")
}
```

We rarely use `len()` on channels. So rarely that I had to look up how exactly it works.

Reply:
Same comment further up in the test.